cupy/sorting/search.py

```diff
@@ -181,7 +182,131 @@ def where(condition, x=None, y=None):
     return _where_ufunc(condition.astype('?'), x, y)
 
 
-# TODO(okuta): Implement searchsorted
+# This is to allow using the same kernels for all dtypes, ints & floats
+# as nan is a special case
+_preamble = '''
+template<typename T>
+__device__ bool _isnan(T val) {
+    return false;
+}
+template<>
+__device__ bool _isnan<float16>(float16 val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<float>(float val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<double>(double val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<const complex<double>&>(const complex<double>& val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<const complex<float>&>(const complex<float>& val) {
+    return isnan(val);
+}
+'''
+
+
+_searchsorted_kernel = core.ElementwiseKernel(
+    'S x, raw T bins, int64 n_bins, bool side',
+    'U y',
+    '''
+    bool greater = (side ? x >= bins[n_bins-1] : x > bins[n_bins-1]);
+    if(_isnan<S>(x) || greater) {
+        y = n_bins;
+        return;
+    }
+    long long left = 0;
+    long long right = n_bins-1;
+    while (left < right) {
+        long long m = left + (right - left) / 2;
+        bool mid = (side ? bins[m] <= x : bins[m] < x);
```

Comment:
Is the name `mid` appropriate?
I think this condition can be directly embedded in the `if` statement.